### PR TITLE
Cache OptimizationResultCollection.to_records calls

### PR DESCRIPTION
**Pull request by ntBre**
The `CachedResultCollection` holds the fields that `MoleculeStore.from_qcsubmit_collection` uses from the `OptimizationResultCollection.to_records` call so that the records don't have to be retrieved from QCArchive each time a `MoleculeStore` is constructed.

I'm definitely not married to the name, so feel free to recommend something you prefer.

`from_cached_result_collection` is copy-pasted from `from_qcsubmit_collection`, with `MoleculeRecord.from_molecule` and `QMConformerRecord.from_qcarchive_record` effectively in-lined to create the objects directly.

### CODE REVIEW

File: ibstore/cached_result.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    qc_record_id: int
+    qc_record_final_energy: float
+
+    def to_dict(self):
+        assert self.coordinates.units == unit.angstrom
+        return dict(
+            mapped_smiles=self.mapped_smiles,
+            inchi_key=self.inchi_key,
+            # this is a pint.Quantity-wrapped numpy array
+            coordinates=self.coordinates.magnitude.tolist(),
```
[the remark this answers is not in the document]
Comment (ntBre) [reply ibstore/cached_result.py:29]: I just pushed my README addition. I couldn't remember exactly where I got the original tiny-opt `OptimizationResultCollection`, so I just took the first 200 entries from the Sage 2.1.0 repo to be sure. I included that subset as `tiny-opt-dataset.json` as well for full reproducibility.

I also tested this script in the `ibstore/_tests/data` directory and for the tiny dataset, the `from_qcsubmit_collection` version took about 5.5 seconds and the cached version about 3.5 seconds on my computer. I ran it three times and also tried swapping the order of the two and got pretty consistent results.

```python
import logging
import os
import time

from openff.qcsubmit.results import OptimizationResultCollection

from ibstore import MoleculeStore
from ibstore.cached_result import CachedResultCollection

logging.getLogger("openff").setLevel(logging.ERROR)


dataset = "tiny-opt-dataset.json"
opt = OptimizationResultCollection.parse_file(dataset)
start = time.time()
MoleculeStore.from_qcsubmit_collection(opt, "foo1.sqlite")
end = time.time()
print(f"from_qcsubmit_collection: {end - start:.1f} sec")

dataset = CachedResultCollection.from_json("tiny-opt.json")
start = time.time()
crc = MoleculeStore.from_cached_result_collection(dataset, "foo2.sqlite")
end = time.time()
print(f"from_cached_result_collection: {end - start:.1f} sec")

os.remove("foo1.sqlite")
os.remove("foo2.sqlite")
```

**Comment by ntBre** [comment]
I just pushed my README addition. I couldn't remember exactly where I got the original tiny-opt `OptimizationResultCollection`, so I just took the first 200 entries from the Sage 2.1.0 repo to be sure. I included that subset as `tiny-opt-dataset.json` as well for full reproducibility.

I also tested this script in the `ibstore/_tests/data` directory and for the tiny dataset, the `from_qcsubmit_collection` version took about 5.5 seconds and the cached version about 3.5 seconds on my computer. I ran it three times and also tried swapping the order of the two and got pretty consistent results.

```python
import logging
import os
import time

from openff.qcsubmit.results import OptimizationResultCollection

from ibstore import MoleculeStore
from ibstore.cached_result import CachedResultCollection

logging.getLogger("openff").setLevel(logging.ERROR)


dataset = "tiny-opt-dataset.json"
opt = OptimizationResultCollection.parse_file(dataset)
start = time.time()
MoleculeStore.from_qcsubmit_collection(opt, "foo1.sqlite")
end = time.time()
print(f"from_qcsubmit_collection: {end - start:.1f} sec")

dataset = CachedResultCollection.from_json("tiny-opt.json")
start = time.time()
crc = MoleculeStore.from_cached_result_collection(dataset, "foo2.sqlite")
end = time.time()
print(f"from_cached_result_collection: {end - start:.1f} sec")

os.remove("foo1.sqlite")
os.remove("foo2.sqlite")
```

**Comment by mattwthompson** [comment]
Does that speedup (~40% faster) hold for larger systems? I recall (poorly, perhaps) much better speedups with something you did. Maybe I'm just confusing this and the rest of #21?

**Review by mattwthompson — APPROVED** [review]
I read through everything again and this is great. I love the simplicity of plucking out what we actually need from the QCSubmit objects.

**Comment by ntBre** [comment]
I was going to guess but couldn't help checking, and the speedup does hold for the industry dataset (72459 records), at least over one run on my computer. `from_qcsubmit_collection` took 2672.4 sec, and the cached version took 1754.1 sec for a ~34% reduction. Still, I think the other changes in #21 make a bigger impact like you said.

**Comment by mattwthompson** [comment]
Some other, rougher, numbers from my own notes:

The dataset of 200 molcules in this PR takes ~14 ms to load from disk and ~140 ms to load into a `MoleculeStore`, numbers that look even better when dividing through by 200 molecules.

**Comment by mattwthompson** [comment]
Any idea why I'm seeing a ~200x speedup?

```python
In [1]: import random

In [2]: from openff.qcsubmit.results import OptimizationResultCollection

In [3]: from ibstore.cached_result import CachedResultCollection

In [4]: from ibstore import MoleculeStore

In [5]: cached = CachedResultCollection.from_json("ibstore/_tests/data/tiny-opt.json")

In [6]: qcs = OptimizationResultCollection.parse_file("ibstore/_tests/data/tiny-opt-dataset.json")

In [7]: %%timeit
   ...: MoleculeStore.from_cached_result_collection(cached, f"{random.random()}.sqlite")
   ...:
   ...:
136 ms ± 1.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %%timeit
   ...: MoleculeStore.from_qcsubmit_collection(qcs, database_name=f"{str(random.random())}.sqlite",)
   ...:
   ...:
32.9 s ± 2.03 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

**Comment by ntBre** [comment]
Not really, that's both faster than I saw on my computer for the cache and slower than I saw for the `to_records` call. The latter could be some kind of network issue? For `from_qcsubmit_collection` using your code above, I just got `4.07 s ± 26.2 ms`. And for the cached version,  I got `1.29 s ± 18.3 ms`, which is still faster than I saw yesterday but at least the same order of magnitude.

**Comment by mattwthompson** [comment]
I'm willing to chalk it up the difference in loading QCSubmit data to network throttling of some sort

The cache loading 10x faster on my machine than your is surprising, but could be down to any number of factors. If there is a difference that's down to hardware, the performance on a cluster is really what matters so I'm willing to call it here

**Pull request merged**